text/0000-decorators.md

```diff
@@ -0,0 +1,1230 @@
+
+```js
+class Person {
+  @alias('prefix') title;
+}
+```
+
+Really, what's going on there is _not_ that we are invoking the `@alias`
+decorator with parameters. We are invoking a function which _returns_ a
+decorator, so it desugars to:
+
+```js
+const aliasForPrefix = alias('prefix');
+
+class Person {
+  @aliasForPrefix title;
+}
+```
+
+Therefore, the `alias` function _must_ itself return a decorator function.
+However, this conflicts with usage in the legacy programming model:
+
+```js
+const Person = EmberObject.extend({
+  title: alias('prefix')
+})
+```
+
+We just established that `alias` must return a decorator function, but here it
+is with the exact same arguments, and it needs to return a `ComputedProperty`
+instance. There is nothing we can branch on here - in both cases, `alias` only
+receives the string `'prefix'`, so it has no context for how it will be used.
+
+The native class piece of this puzzle is completely inflexible. A decorator must
+be a function, there is no choice about it. However, the _Ember_ piece is _very_
+flexible. The legacy object model just needs a way to get the meta information
+for the property when the class is being finalized. We can either assign the
+meta information to the decorator function directly, or we can associate it via
+a `WeakMap`.
+
+The benefit of doing this is that the entire Ember ecosystem will get decorator
+support with no extra work required. Because standard `computed` definitions
+will work as decorators, existing macros will also work as decorators with no
+changes to existing code.
+
+All of Ember's built in computed macros in the `@ember/object/computed` module
+will also become decorators with no extra work. However, the injection macros
+will require slight updates as they use a subclass of the `ComputedProperty`
+class. These updates should be relatively minor, and will follow the same
+strategy as `computed()`.
+
+#### Usage and API
+
+The API for `computed` will remain mostly the same. The key differences will be:
+
+1. The result of `computed` will be a decorator which can be applied directly to
+   native _getters_, _setters_, and _class fields_.
+2. The `ComputedPropertyConfig` (the getter/setter functions) argument provided
+   to `computed` will now be optional when used as a decorator on a native
+   getter or setter, and the native getter/setter will be used instead.
+
+```ts
+function computed(...args: (string | ComputedPropertyConfig)[]): PropertyDecorator;
+```
+
+The function signatures of all existing macros, including `inject` macros, would
+change in the same way.
+
+In general usage, these three definitions are equivalent:
+
+```js
+import EmberObject, { computed } from '@ember/object';
+
+const Person = EmberObject.extend({
+  fullName: computed('firstName', 'lastName', {
+    get() {
+      return `${this.firstName} ${this.lastName}`;
+    }
+  })
+});
+
+class Person {
+  @computed('firstName', 'lastName')
```

Comment:
Major selling point that you may want to point out:

If `@computed` works in the same way as the current ember-decorators implementation, it may be used on a vanilla ES6 class _without having to extend from `EmberObject`_.